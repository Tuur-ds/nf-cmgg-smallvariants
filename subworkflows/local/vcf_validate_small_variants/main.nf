include { RTGTOOLS_VCFEVAL      } from '../../../modules/nf-core/rtgtools/vcfeval/main'
include { RTGTOOLS_ROCPLOT      } from '../../../modules/nf-core/rtgtools/rocplot/main'

workflow VCF_VALIDATE_SMALL_VARIANTS {

    take:
    ch_vcf                          // [mandatory] channel: [ meta, vcf, tbi, truth_vcf, truth_tbi ]
    ch_beds                         // [mandatory] channel: [ meta, truth_bed, region_bed ]
    ch_vcfeval_sdf                  // [vcfeval only] channel: [ meta, sdf ]

    main:

    def ch_versions = Channel.empty()
    def ch_input = ch_vcf.join(ch_beds, failOnDuplicate: true, failOnMismatch: true)

    RTGTOOLS_VCFEVAL(
        ch_input,
        ch_vcfeval_sdf
    )
    ch_versions = ch_versions.mix(RTGTOOLS_VCFEVAL.out.versions.first())

    def ch_rocplot_input = RTGTOOLS_VCFEVAL.out.snp_roc
        .map { meta, tsv ->
            [ meta + [roc_type:'snp'], tsv ]
        }
        .mix(
            RTGTOOLS_VCFEVAL.out.non_snp_roc.map { meta, tsv ->
                [ meta + [roc_type:'non_snp'], tsv ]
            },
            RTGTOOLS_VCFEVAL.out.weighted_roc.map { meta, tsv ->
                [ meta + [roc_type:'weighted'], tsv ]
            }
        )

    def ch_vcfeval_true_positive_vcf               = RTGTOOLS_VCFEVAL.out.tp_vcf
    def ch_vcfeval_true_positive_vcf_tbi           = RTGTOOLS_VCFEVAL.out.tp_tbi
    def ch_vcfeval_false_negative_vcf              = RTGTOOLS_VCFEVAL.out.fn_vcf
    def ch_vcfeval_false_negative_vcf_tbi          = RTGTOOLS_VCFEVAL.out.fn_tbi
    def ch_vcfeval_false_positive_vcf              = RTGTOOLS_VCFEVAL.out.fp_vcf
    def ch_vcfeval_false_positive_vcf_tbi          = RTGTOOLS_VCFEVAL.out.fp_tbi
    def ch_vcfeval_true_positive_baseline_vcf      = RTGTOOLS_VCFEVAL.out.baseline_vcf
    def ch_vcfeval_true_positive_baseline_vcf_tbi  = RTGTOOLS_VCFEVAL.out.baseline_tbi
    def ch_vcfeval_summary                         = RTGTOOLS_VCFEVAL.out.summary
    def ch_vcfeval_phasing                         = RTGTOOLS_VCFEVAL.out.phasing
    def ch_vcfeval_snp_roc                         = RTGTOOLS_VCFEVAL.out.snp_roc
    def ch_vcfeval_non_snp_roc                     = RTGTOOLS_VCFEVAL.out.non_snp_roc
    def ch_vcfeval_weighted_roc                    = RTGTOOLS_VCFEVAL.out.weighted_roc

    RTGTOOLS_ROCPLOT(
        ch_rocplot_input
    )

    ch_versions = ch_versions.mix(RTGTOOLS_ROCPLOT.out.versions.first())

    def rocplot_out_png = RTGTOOLS_ROCPLOT.out.png
        .branch { meta, _png ->
            def roc_type = meta.roc_type

            snp:        roc_type == "snp"
            non_snp:    roc_type == "non_snp"
            weighted:   roc_type == "weighted"
        }

    def rocplot_out_svg = RTGTOOLS_ROCPLOT.out.svg
        .branch { meta, _svg ->
            def roc_type = meta.roc_type

            snp:        roc_type == "snp"
            non_snp:    roc_type == "non_snp"
            weighted:   roc_type == "weighted"
        }

    def ch_rtgtools_snp_png_rocplot        = rocplot_out_png.snp
    def ch_rtgtools_non_snp_png_rocplot    = rocplot_out_png.non_snp
    def ch_rtgtools_weighted_png_rocplot   = rocplot_out_png.weighted

    def ch_rtgtools_snp_svg_rocplot        = rocplot_out_svg.snp
    def ch_rtgtools_non_snp_svg_rocplot    = rocplot_out_svg.non_snp
    def ch_rtgtools_weighted_svg_rocplot   = rocplot_out_svg.weighted

    emit:
    vcfeval_true_positive_vcf               = ch_vcfeval_true_positive_vcf               // channel: [ meta, vcf ]
    vcfeval_true_positive_vcf_tbi           = ch_vcfeval_true_positive_vcf_tbi           // channel: [ meta, tbi ]
    vcfeval_false_negative_vcf              = ch_vcfeval_false_negative_vcf              // channel: [ meta, vcf ]
    vcfeval_false_negative_vcf_tbi          = ch_vcfeval_false_negative_vcf_tbi          // channel: [ meta, tbi ]
    vcfeval_false_positive_vcf              = ch_vcfeval_false_positive_vcf              // channel: [ meta, vcf ]
    vcfeval_false_positive_vcf_tbi          = ch_vcfeval_false_positive_vcf_tbi          // channel: [ meta, tbi ]
    vcfeval_true_positive_baseline_vcf      = ch_vcfeval_true_positive_baseline_vcf      // channel: [ meta, vcf ]
    vcfeval_true_positive_baseline_vcf_tbi  = ch_vcfeval_true_positive_baseline_vcf_tbi  // channel: [ meta, tbi ]
    vcfeval_summary                         = ch_vcfeval_summary                         // channel: [ meta, summary ]
    vcfeval_phasing                         = ch_vcfeval_phasing                         // channel: [ meta, phasing ]
    vcfeval_snp_roc                         = ch_vcfeval_snp_roc                         // channel: [ meta, tsv ]
    vcfeval_non_snp_roc                     = ch_vcfeval_non_snp_roc                     // channel: [ meta, tsv ]
    vcfeval_weighted_roc                    = ch_vcfeval_weighted_roc                    // channel: [ meta, tsv ]

    rtgtools_snp_png_rocplot                = ch_rtgtools_snp_png_rocplot                // channel: [ meta, png ]
    rtgtools_non_snp_png_rocplot            = ch_rtgtools_non_snp_png_rocplot            // channel: [ meta, png ]
    rtgtools_weighted_png_rocplot           = ch_rtgtools_weighted_png_rocplot           // channel: [ meta, png ]
    rtgtools_snp_svg_rocplot                = ch_rtgtools_snp_svg_rocplot                // channel: [ meta, svg ]
    rtgtools_non_snp_svg_rocplot            = ch_rtgtools_non_snp_svg_rocplot            // channel: [ meta, svg ]
    rtgtools_weighted_svg_rocplot           = ch_rtgtools_weighted_svg_rocplot           // channel: [ meta, svg ]

    versions = ch_versions                  // channel: [ versions.yml ]
}

