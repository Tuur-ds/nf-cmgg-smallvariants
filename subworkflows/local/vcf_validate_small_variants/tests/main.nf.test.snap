{
    "default": {
        "content": [
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.tp.vcf.gz:md5,0449f4185b63d0b988b7f9de6a6b2fb3"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.tp.vcf.gz.tbi:md5,637a1a87502b3764ad22017b16ccb3ac"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.fn.vcf.gz:md5,688c27c6785313c1a0ac9fdcf430e0eb"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.fn.vcf.gz.tbi:md5,092a7a3162e7cff25d273525751eb284"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.fp.vcf.gz:md5,d11b30ab764bd9c3f8294b51043a3e1a"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.fp.vcf.gz.tbi:md5,092a7a3162e7cff25d273525751eb284"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.tp-baseline.vcf.gz:md5,748b2b41e2f5916fd0e5fbc2aea60d3d"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.tp-baseline.vcf.gz.tbi:md5,8c1a12430a7714b000c11e471175d6e1"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.summary.txt:md5,11416903ca828b22bd64cf297f2db999"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.phasing.txt:md5,dfcf0fc6ad7e98c737a7a0afaf2c2733"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.snp_roc.tsv.gz:md5,306d33ea1cf1d797d0f52234d2ff901e"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.non_snp_roc.tsv.gz:md5,31e6719897dd4f7d3d0a9cb4943693cf"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385"
                    },
                    "NA24143.weighted_roc.tsv.gz:md5,e1e87f61f55f976162be76fe993423e1"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385",
                        "roc_type": "snp"
                    },
                    "NA24143.snp.png"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385",
                        "roc_type": "non_snp"
                    },
                    "NA24143.non_snp.png"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385",
                        "roc_type": "weighted"
                    },
                    "NA24143.weighted.png"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385",
                        "roc_type": "snp"
                    },
                    "NA24143.snp.svg"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385",
                        "roc_type": "non_snp"
                    },
                    "NA24143.non_snp.svg"
                ]
            ],
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143,NA24149,NA24385",
                        "roc_type": "weighted"
                    },
                    "NA24143.weighted.svg"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-08T15:56:00.946806992"
    }
}